cpp/src/arrow/compute/kernels/cast.cc

```diff
@@ -1009,6 +1039,53 @@ struct CastFunctor<TimestampType, StringType> {
   }
 };
 
+// ----------------------------------------------------------------------
+// Binary to String
+//
+
+template <typename I>
+struct CastFunctor<
+    StringType, I,
+    typename std::enable_if<is_binary_to_string<StringType, I>::value>::type> {
+  void operator()(FunctionContext* ctx, const CastOptions& options,
+                  const ArrayData& input, ArrayData* output) {
+    BinaryArray binary(input.Copy());
+
+    if (options.allow_invalid_utf8) {
+      ZeroCopyData(input, output);
+      return;
+    }
+
+    // TODO(fsaintjacques): This is not the most optimal place to put this.
+    util::InitializeUTF8();
+
```

Comment:
Note there is an interesting potential optimization here: if the whole binary array is ASCII, then regardless of the exact offsets it is also valid UTF8. It's a computationally cheap check.

(we could also imagine doing this chunk-wise, if we want the optimization to apply to partially ASCII arrays, and if we want to avoid two full sequentiel reads if the data is large and non-ASCII)

Reply:
Wouldn't that be equivalent to pass the `allow_invalid_utf8=true` flag? Maybe we should rename the option, such that if the user knows the input is valid, there's no need to do validation.

Reply:
I'd propose `skip_utf8_validation`, it preserve the same boolean-ness of `safe` negation like the other flags, and it cover both cases while leaving no place to interpretation.

Reply:
I see your point, but that would imply doing 2 pass on the array, that's... potentially very costly since we're likely memory bound.

Reply:
I'm not sure I understand your first two comments in relation to the comment I posted :-)

Reply:
As for the third-point, yes, doing it chunk-wise may be more robust (e.g. process 1000 offsets at once).

Reply:
I understood that the user _knew_ that it's valid ASCII payload.